Review and Add the Guidelines for Writing Ballerina By Examples (BBEs) (#16456)
ballerina-platform/ballerina-lang: docs/bbe-guidelines/bbe-guidelines.md

# Guidelines for Writing Ballerina By Examples (BBEs)

- [General best practices](#general-best-practices)
- [Practices related to the code comments](#practices-related-to-the-code-comments)
- [Practices related to the folder structure](#practices-related-to-the-folder-structure)

## General best practices

1. Format the Ballerina source using an IDE Plugin.

2. Always, think about the length of the example. A good example should be short.

3. Make the examples self-contained (i.e., avoid depending on concepts other than the one being illustrated) as much as possible.

> 4. As a practice, use `ballerina/io` methods in main examples and `ballerina/log` in services examples. Do not use both `io:println` and `log:printInfo` in the same example.

Reviewer: ```suggestion
4. As a practice, use `ballerina/io` methods in main examples and `ballerina/log` methods in services examples. E.g., Do not use both `io:println` and `log:printInfo` in the same example.
```

> 5. It is recommended to limit the maximum character count per line to 80 in .bal files. Otherwise, the lines get automatically wrapped in the [Ballerina website](https://ballerina.io/) reducing the readability as shown below.

Reviewer: ```suggestion
5. It is recommended to limit the maximum character count per line to 80 in `.bal` files. Otherwise, the lines get automatically wrapped in the [Ballerina website](https://ballerina.io/) reducing the readability as shown below.
```

    Example;

    <img src="images/line-wrap.png" width="600" height="70" alt="wrapping lines" />

6. All keywords and any other word, which needs to be highlighted should be added within backquotes (e.g., `xml`). Do not use a single quote as it will not get highlighted. 

    Example;

    - In the code:

        <img src="images/backquoted-keywords.png" width="400" height="58" alt="backquoted keywords" />

    - In the Baleerina website:

        <img src="images/keyword-in-bio.png" width="400" height="80" alt="keywords in bio" />

    > **Note:** However, the context should be taken into account when formatting words/phrases with backquotes. For example, although the word "record" is a keyword of the language, it need not be highlighted if you are referring to the record construct of the language as shown below.

>     ````

Reviewer: ```suggestion
    ```ballerina
```

Author: Fixed.

    // Defines an open record type named Student. The { and } delimiters indicate that in addition to the defined fields,
    // this record type allows additional fields with pure-typed (i.e., anydata|error) values.
    // The descriptor record { } is equivalent to record {| (anydata|error)...; |}.
    type Student record {

>     string name;

Reviewer 2: ```suggestion
        string name;
```

>     int age;

Reviewer 2: ```suggestion
        int age;
```

>     Grades grades;

Reviewer 2: ```suggestion
        Grades grades;
```

    };

>     ````

Reviewer: ```suggestion
    ```
```

7. Remove unused imports in `.bal` files.

> 8. If a new example is added/deleted, update the [index.json](https://github.com/ballerina-platform/ballerina-lang/blob/master/examples/index.json) file as well.

Reviewer: ```suggestion
8. If a new example is added or an existing one is deleted/renamed, update the [index.json](https://github.com/ballerina-platform/ballerina-lang/blob/master/examples/index.json) file as well.
```

## Practices related to the code comments

> 1. Add comments to the code blocks as much as possible by starting them with “//”. They will be moved to the RHS boxes in the [Ballerina website](https://ballerina.io/) as a mechanism to describe the code.

Reviewer: ```suggestion
1. Add comments to the code blocks as much as possible by starting them with “//”. They will be moved to the right hand side boxes in the [Ballerina By Example](https://ballerina.io/learn/by-example/) as a mechanism to describe the code.
```

    Example;

    - In the code:

        <img src="images/code-comments.png" width="600" height="80" alt="adding code comments" />

    - In the [Ballerina website](https://ballerina.io/):

        ![comment boxes](images/comment-boxes.png "comment-boxes")

> 2. Since comments are used in the RHS as a mechanism to describe the code, they should be valid sentences (i.e., start with an upper case letter and end with a full stop etc.).

Reviewer: ```suggestion
2. Since comments are used in the right hand side boxes as a mechanism to describe the code, they should be valid sentences (i.e., start with an upper case letter and end with a full stop etc.).
```

3. After a comment is strated, it will continue being applied to the subsequent lines in the file until another comment or an empty line is found. Therefore, use comments/new lines appropriately to ensure that they apply only to the relevant lines.

> 4. There is no restriction on the max number of chars that should be there in comment lines as they get wrapped automatically in the RHS side in the [Ballerina website](https://ballerina.io/). However, since users can refer the same in the code in GitHub, it is better if we can have the same char limit as of a code line (i.e., 80) to increase readability.

Reviewer: ```suggestion
4. There is no restriction on the max number of characters that should be there in comment lines as they get wrapped automatically in the right hand side in the [Ballerina By Example](https://ballerina.io/learn/by-example/). However, since users can refer the same in the code in GitHub, it is better if we can have the same character limit as of a code line (i.e., 80) to increase readability.
```

    For example, it will not be readable if comments are significantly longer than the code line as shown below. 

    ![long comments](images/long-comments.png "long-comments")

## Practices related to the folder structure

> 1. Each BBE should be in a separate directory with the example name. 

Reviewer: Add the github repo url of the examples directory.

Author: Fixed.

2. Directory name should be in all lowercase letters with words separated by “-” (e.g., abstract-objects).

3. Each example should contain at least the following files. 

    > **Tip:** File names should be the same as folder name, with the “-” being replaced with the underscore (e.g., abstract_objects).

>     .bal - The code of the example to be displayed in the [Ballerina website](https://ballerina.io/).

Reviewer: ```suggestion
    `.bal` - The code of the example to be displayed in the [Ballerina By Examples](https://ballerina.io/learn/by-example/).
```

>     .description - The description of the example displayed at the top of each example in the [Ballerina website](https://ballerina.io/).

Reviewer: ```suggestion
    `.description` - The description of the example displayed at the top of each example in the [Ballerina By Example](https://ballerina.io/learn/by-example/).
```

>     .out - Output of the example displayed at the bottom black colour box in the [Ballerina website](https://ballerina.io/).

Reviewer: ```suggestion
    `.out` - Output of the example displayed at the bottom black colour box in the [Ballerina By Example](https://ballerina.io/learn/by-example/).
```

>     _test.bal - Contains the test to validate the output of the BBE during the build time. 

Reviewer: ```suggestion
    `<example_name>_test.bal` - Contains the test to validate the output of the BBE during the build time. 
```

>     <img src="images/bbe-folder-structure.png" width="350" height="200" alt="BBE folder structure" />

Reviewer: This image seems blurred

4. As a common pattern, use the following format for `.out` files and customize it only when required (e.g., to add more command line args etc.).

- For an example with main;

    ```
    # To run this sample, navigate to the directory that contains the

>     # `.bal` file and issue the `ballerina run` command.

Reviewer: ```suggestion
    # `.bal` file and execute the `ballerina run` command.
```

    $ ballerina run <sample_file_name>.bal
    ```
- For an example with a service;

    ```
    # To start the service, navigate to the directory that contains the

>     # `.bal` file and issue the `ballerina run` command.

Reviewer: ```suggestion
    # `.bal` file and execute the `ballerina run` command.
```

    $ ballerina run hello_world_service.bal
    ```

5. Service examples need two `.out` files. One (i.e., the `server.out` file) to display the server output and the other (i.e., the `client.out` file) to display the cURL command and the output.

    <img src="images/service-example-structure.png" width="400" height="220" alt="service example structure" />

> 6. Unless it is really required, it is not encouraged to have multiple .bal files in the same example. In case if multiple are needed, each BAL file can have its own name and the respective `.out` files should match the name of the respective `.bal` file. 

Reviewer: ```suggestion
6. Unless it is really required, it is not encouraged to have multiple `.bal` files in the same example. In case if multiple are needed, each BAL file can have its own name and the respective `.out` files should match the name of the respective `.bal` file. 
```

6. All keywords and any other word, which needs to be highlighted should be added within backquotes (e.g., `xml`). Do not use a single quote as it will not get highlighted. 

    Example;

    - In the code:

        <img src="images/backquoted-keywords.png" width="400" height="58" alt="backquoted keywords" />

    - In the Baleerina website:

        <img src="images/keyword-in-bio.png" width="400" height="80" alt="keywords in bio" />

    > **Note:** However, the context should be taken into account when formatting words/phrases with backquotes. For example, although the word "record" is a keyword of the language, it need not be highlighted if you are referring to the record construct of the language as shown below.

    ````
    // Defines an open record type named Student. The { and } delimiters indicate that in addition to the defined fields,
    // this record type allows additional fields with pure-typed (i.e., anydata|error) values.
    // The descriptor record { } is equivalent to record {| (anydata|error)...; |}.
    type Student record {
        string name;
        int age;
    Grades grades;
    };

>     ```

Reviewer: Seems the above content is not properly visible in the page. End ``` is also visible and horizontal scroll bar is added.

Author: Fixed.

7. Remove unused imports in `.bal` files.

8. If a new example is added or an existing one is deleted/renamed, update the [index.json](https://github.com/ballerina-platform/ballerina-lang/blob/master/examples/index.json) file as well.

## Practices related to the code comments

1. Add comments to the code blocks as much as possible by starting them with “//”. They will be moved to the right hand side boxes in the [Ballerina By Example](https://ballerina.io/learn/by-example/) as a mechanism to describe the code.

    Example;

    - In the code:

        <img src="images/code-comments.png" width="600" height="80" alt="adding code comments" />

    - In the [Ballerina website](https://ballerina.io/):

        ![comment boxes](images/comment-boxes.png "comment-boxes")

> 2. Since comments are used in the right hand side boxes as a mechanism to describe the code, they should be valid sentences (i.e., start with an upper case letter and end with a full stop etc.).

Reviewer: "as a mechanism to describe the code" is also there at the end of the 1st point. IMO, that duplication is not good.

Author: Fixed as "Since comments are used to describe the code,".

3. After a comment is strated, it will continue being applied to the subsequent lines in the file until another comment or an empty line is found. Therefore, use comments/new lines appropriately to ensure that they apply only to the relevant lines.

4. There is no restriction on the max number of characters that should be there in comment lines as they get wrapped automatically in the right hand side in the [Ballerina By Example](https://ballerina.io/learn/by-example/). However, since users can refer the same in the code in GitHub, it is better if we can have the same character limit as of a code line (i.e., 80) to increase readability.

    For example, it will not be readable if comments are significantly longer than the code line as shown below. 

    ![long comments](images/long-comments.png "long-comments")

## Practices related to the folder structure

> 1. Each BBE should be in a separate directory in the [ballerina-lang Git Hub repository](https://github.com/ballerina-platform/ballerina-lang/tree/master/examples) named by the example (e.g., https://github.com/ballerina-platform/ballerina-lang/tree/master/examples/artemis-transaction-producer).

Reviewer: ```suggestion
1. Each BBE should be in a separate directory in the [ballerina-lang GitHub repository](https://github.com/ballerina-platform/ballerina-lang/tree/master/examples) named by the example (e.g., https://github.com/ballerina-platform/ballerina-lang/tree/master/examples/artemis-transaction-producer).
```

Reviewer: Also shall we use a more common example such as hello-world-service as the example link?

Author: Fixed "GitHub" and added [1] as the link.

[1] https://github.com/ballerina-platform/ballerina-lang/tree/master/examples/hello-world

2. Directory name should be in all lowercase letters with words separated by “-” (e.g., abstract-objects).

3. Each example should contain at least the following files. 

    > **Tip:** File names should be the same as folder name, with the “-” being replaced with the underscore (e.g., abstract_objects).

    `.bal` - The code of the example to be displayed in the [Ballerina By Example](https://ballerina.io/learn/by-example/).

    `.description` - The description of the example displayed at the top of each example in the [Ballerina By Example](https://ballerina.io/learn/by-example/).

    `.out` - Output of the example displayed at the bottom black colour box in the [Ballerina By Example](https://ballerina.io/learn/by-example/).

>     `<example_name>_test.bal` - Contains the test to validate the output of the BBE during the build time. 

Reviewer: ```suggestion
    `<example_name>_test.bal` - Contains the tests to validate the output of the BBE during the build time. 
```

Author: Fixed.

    <img src="images/bbe-folder-structure.png" width="300" height="150" alt="BBE folder structure" />

4. As a common pattern, use the following format for `.out` files and customize it only when required (e.g., to add more command line args etc.).

- For an example with main;

    ```
    # To run this sample, navigate to the directory that contains the
    # `.bal` file and execute the `ballerina run` command.
    $ ballerina run <sample_file_name>.bal
    ```
- For an example with a service;

    ```
    # To start the service, navigate to the directory that contains the
    # `.bal` file and execute the `ballerina run` command.
    $ ballerina run hello_world_service.bal
    ```

5. Service examples need two `.out` files. One (i.e., the `server.out` file) to display the server output and the other (i.e., the `client.out` file) to display the cURL command and the output.

    <img src="images/service-example-structure.png" width="330" height="180" alt="service example structure" />

> 6. Unless it is really required, it is not encouraged to have multiple `.bal` files in the same example. In case if multiple are needed, each BAL file can have its own name and the respective `.out` files should match the name of the respective `.bal` file. 

Reviewer: ```suggestion
6. Unless it is really required, it is not encouraged to have multiple `.bal` files in the same example. In case if multiple are needed, each `.bal` file can have its own name and the respective `.out` files should match the name of the respective `.bal` file. 
```

Author: Fixed.

# Guidelines for Writing Ballerina By Examples (BBEs)

## General Best Practices

> 1. The maximum character count per line should be 80 in .bal files. Otherwise, the lines will be wrapped in the [Ballerina](https://ballerina.io/) website reducing the readability.

Reviewer 2: Was the limit 80? I recall it being 60 something.

Reviewer 2: And the character count doesn't apply for comments in the .bal file, though it would be a good practice to keep it to 80 if possible

Reviewer 2: Ah ignore the 2nd comment above. Can see that it's already addressed

Author: Counted from the below line in [1]. The character limit is 80. Reworded as follows:

"It is recommended to limit the maximum character count per line to 80 in .bal files. Otherwise, the lines get automatically wrapped in the [Ballerina website](https://ballerina.io/) reducing the readability as shown below."

"io:println("Insert to table failed: " + <string>retWithKey.detail().message);"

[1] https://github.com/ballerinax/jdbc/blob/master/jdbc-extension-examples/examples/jdbc-client-crud-operations/jdbc_client_crud_operations.bal

Example:

<p id="gdcalert1" ><span style="color: red; font-weight: bold">>>>>>  gd2md-html alert: inline image link here (to images/Guidelines-in0.png). Store image on your image server and adjust path/filename if necessary. </span><br>(<a href="#">Back to top</a>)(<a href="#gdcalert2">Next alert</a>)<br><span style="color: red; font-weight: bold">>>>>> </span></p>

![alt_text](images/Guidelines-in0.png "image_tooltip")

> *   **Add comments to the code blocks as much as possible with “//”. They will move to the RHS box in b.io. So comments are used as a mechanism to describe the code.**

Reviewer 2: Who's the intended audience of this doc? If this is applicable to everyone, then we should probably minimize the use of abbreviations such as "b.io"

Author: Changed as "[Ballerina website](https://ballerina.io/)".

![alt_text](images/Guidelines-in2.png "image_tooltip")

*   **Since comments are used in RHS they should be a valid sentence. (i.e. Start with upper case letter, ends with full stop etc)**
*   **A comment applies to the subsequent lines in the file until another comment or an empty line is found. Use comments/new lines appropriately to ensure that it applies only to the relevant lines.**
*   **No restriction for the max chars in comment lines as they will go to RHS side and wraps automatically. But since users can refer the code in GitHub, it is better if we can have the same char limit as the code line (i.e. 80) as it increases the readability of the code file.**

Ex: Comments are significantly longer than the code line, which is not readable. 

<p id="gdcalert4" ><span style="color: red; font-weight: bold">>>>>>  gd2md-html alert: inline image link here (to images/Guidelines-in3.png). Store image on your image server and adjust path/filename if necessary. </span><br>(<a href="#">Back to top</a>)(<a href="#gdcalert5">Next alert</a>)<br><span style="color: red; font-weight: bold">>>>>> </span></p>

![alt_text](images/Guidelines-in3.png "image_tooltip")

> *   **All keywords and any other word which needs to be highlighted should be used with backquotes. (ex:  `xml`).  Don’t use a single quote as it won’t get highlighted. **

Reviewer 2: Formatting words/phrases as code should be done taking the context in to account. We should not be blindly formatting for the sake of adhering to this guideline. For example take the word "record" or "object". These are keywords in the language. But whether to format them as code or not depends on the context. If you are simply referring to the record construct in the language, it should not be highlighted. For example consider the below excerpt from the records BBE. 

> // Defines an open record type named `Student`. The `{` and `}` delimiters indicate that in addition to the defined fields,
// this record type allows additional fields with pure-typed (i.e., `anydata|error`) values.
// The descriptor `record { }` is equivalent to `record {| (anydata|error)...; |}`.
type Student record {
    string name;
    int age;
    Grades grades;
};

Note how record is not highlighted when the context in which it appears is not code.

Author: Added the above as a note.

![alt_text](images/Guidelines-in5.png "image_tooltip")

*   **We use a common pattern in `.out` files. Use the following format and customize only when necessary (ex: When needed to add more command line args etc)**

<span style="text-decoration:underline;">For an example with main:</span>

# To run this sample, navigate to the directory that contains the

# `.bal` file and issue the `ballerina run` command.

$ ballerina run <sample_file_name>.bal

<span style="text-decoration:underline;">For an example with a service:</span>

# To start the service, navigate to the directory that contains the

# `.bal` file and issue the `ballerina run` command.

$ ballerina run hello_world_service.bal

*   **As a practice, we use `<code>ballerina/io`</code> methods in main examples and `<code>ballerina/log</code>` in services examples. Don’t use both `<code>io:println</code>` and `<code>log:printInfo</code>` in the same sample.</strong>
*   <strong>Remove unused imports in `.bal` files.</strong>

> *   <strong>Format the ballerina source using an IDE Plugin. </strong>

Reviewer 2: ```suggestion
*   <strong>Format the Ballerina source using an IDE Plugin. </strong>
```

Author: Fixed.

> *   <strong>Always think about the length of the sample. A good example should be short.</strong>

Reviewer 2: And it should be self-contained as much as possible. i.e., should avoid depending on concepts other than the one being illustrated  as much as possible.

Author: Added this as follows under "General best practices".

"Make the examples self-contained (i.e., avoid depending on concepts other than the one being illustrated) as much as possible."

*   <strong>If a new example is added/deleted, update the [index.json](https://github.com/ballerina-platform/ballerina-lang/blob/master/examples/index.json) file as well.</strong>

**<span style="text-decoration:underline;">Folder Structure</span>**

*   **Each BBE is in a separate directory with the sample name. Folder name should be all lowercase and words are separated by “-”. (ex: abstract-objects)**

> *   **Each example should contains at least following files. File name should be same as folder name, but “-” is replaced with underscore. (ex: abstract_objects)**

Reviewer 2: ```suggestion
*   **Each example should contain at least following files. File name should be same as folder name, but “-” is replaced with underscore. (ex: abstract_objects)**
```

Author: fixed.

<p id="gdcalert7" ><span style="color: red; font-weight: bold">>>>>>  gd2md-html alert: inline image link here (to images/Guidelines-in6.png). Store image on your image server and adjust path/filename if necessary. </span><br>(<a href="#">Back to top</a>)(<a href="#gdcalert8">Next alert</a>)<br><span style="color: red; font-weight: bold">>>>>> </span></p>

![alt_text](images/Guidelines-in6.png "image_tooltip")

>            .bal - Sample code to display in b.io.

Reviewer 2: Same comment as above, regarding the use of abbreviations

Author: Changed to "[Ballerina website](https://ballerina.io/)".

           .description - The sample description displayed at the top of each example in b.io

           .out - Output of the sample displayed at the bottom black colour box in b.io.

           _test.bal - Contains the test to validate the output of the BBE during the build time. 

> *   **For service examples, we need to have two .out files. One to display the server output (.server.out file) and the other (.client.out file)  to display the curl command and the output.**

Author: Fixed as "cURL".
